### PR TITLE
describe new backtick string format

### PR DESCRIPTION
**Pull request by rgerhards**
... and also the others, as they were not yet described

see also https://github.com/rsyslog/rsyslog/pull/2419

### CODE REVIEW

File: source/rainerscript/constant_strings.rst [hunk 1]
```diff
@@ -13,3 +13,36 @@ If in doubt how to properly escape, use the `RainerScript String Escape
 Online
 Tool <http://www.rsyslog.com/rainerscript-constant-string-escaper/>`_.
 
+Type of String Constants
```
Comment (deoren) [inline source/rainerscript/constant_strings.rst]: Would this be "Types of String Constants" instead?

Some other ideas (none of which are blockers):

1. Rename file to `string_constants.rst`
1. Update doc title to `String Constants`
1. Add `Uses` section with most of the info from the first section
1. Rename `Type of String Constants` to `Types`

I've push a branch to my repo that will build shortly for demo purposes.

File: source/rainerscript/constant_strings.rst [hunk 2]
```diff
@@ -13,3 +13,36 @@ If in doubt how to properly escape, use the `RainerScript String Escape
...
+
+- single quotes
+
+  Values are used unaltered, execept for escape sequences, which are
+  escaped.
+
+- double quotes
+
+  Right now, equivalent to single quotes, but $ signs need to be escaped.
+  The idea is to support environment variables just like the shell does
```
Comment (deoren) [inline source/rainerscript/constant_strings.rst]: What happens if `$` is not escaped?

**Comment by deoren** [comment]
This is good stuff. I can already see where the new functionality will come in handy and having the existing behavior documented is also good. I've made a few in-line comments for consideration. I've also pushed a branch to my repo with some prototype changes for review. It should start building soon and I'll link to it here once it is ready.

**Comment by deoren** [comment]
Link to live prototype:

http://rsyslog.whyaskwhy.org/pr512-prototype-string-contants/rainerscript/string_constants.html

**Comment by rgerhards** [comment]
updated to now also include ``` `cat filename` ``` -- thought we should add this in the first shot ;-)

File: source/rainerscript/constant_strings.rst [hunk 3]
```diff
@@ -1,15 +1,62 @@
...
+by the shell:
+
+- single quotes
+
+  Values are used unaltered, execept for escape sequences, which are
+  escaped.
+
+- double quotes
+
+  Right now, equivalent to single quotes, but $ signs need to be escaped.
```
Comment (deoren) [inline source/rainerscript/constant_strings.rst]: What happens if `$` is not escaped?

**Comment by deoren** [comment]
> @rgerhards: thx for the feedback!
>
> Updated with all except the file name. I think we should keep it stable until we (hopefully soon) integrate the ./rainerscript files into the main configuration section. Do not confuse Google twice ;-)

Alright. I see that you're referring to #489. If the `./rainerscript` files are merged into the main section then we could update the filename then as needed to match the surrounding content.

> @rgerhards: updated to now also include `cat filename` -- thought we should add this in the first shot ;-)

PRs rsyslog/rsyslog#2426 and rsyslog/rsyslog#2419 are game changers for a lot of configurations. I'm really looking forward to 8.33.0 now. :)

**Comment by deoren** [comment]
Just in case it dropped out of sight, I had one question:

> Right now, equivalent to single quotes, but $ signs need to be escaped

What happens if the user does not escape the `$` sign?

**Comment by rgerhards** [comment]
I thought I added the info on not escaped?

File: source/rainerscript/constant_strings.rst (hunk 3 above)
Comment (rgerhards) [inline source/rainerscript/constant_strings.rst]: Yup, next line 😛😉

Comment (deoren) [reply source/rainerscript/constant_strings.rst]: Evidently I'm blind, I see it now:

> If not escaped, a syntax error will be generated and rsyslog startup be aborted in most cases.

 Thanks!

**Comment by deoren** [comment]
Aside from the filename (which you already indicated would be handled on #489, it looks like this is ready to be merged.

File: source/rainerscript/constant_strings.rst [hunk 4]
```diff
@@ -1,15 +1,62 @@
...
+values that never change during rsyslog's run.
+
+Uses
+----
 String constants are necessary in many places: comparisons,
 configuration parameter values and function arguments, to name a few
 important ones.
 
-In constant strings, special characters are escape by prepending a
+In string constants, special characters are escape by prepending a
```
Comment (deoren) [inline source/rainerscript/constant_strings.rst]: I missed this in the last review:

```diff
-special characters are escape
+special characters are escaped
```

File: source/rainerscript/constant_strings.rst [hunk 5]
```diff
@@ -1,15 +1,62 @@
...
 programming language or PHP.
 
 If in doubt how to properly escape, use the `RainerScript String Escape
 Online
 Tool <http://www.rsyslog.com/rainerscript-constant-string-escaper/>`_.
 
+Types
+-----
+
+Rsyslog provides different types of string constant, closely inspired
```
Comment (deoren) [inline source/rainerscript/constant_strings.rst]:
```diff
-different types of string constant
+different types of string constants
```

**Comment by rgerhards** [comment]
thx, fixed -- please merge if you think it is ok now

**Comment by davidelang** [comment]
I'm assuming that this evaluation only happens at startup? This should be documentation.

**Comment by deoren** [comment]
> @davidelang: I'm assuming that this evaluation only happens at startup? This should be documentation.

I went ahead and created a new issue for this (too easy to quickly glance at remarks on closed issues and then forget to circle back and complete the work). Perhaps not everyone has that problem, but I get distracted easily. ;)

**Comment by rgerhards** [comment]
@davidelang just for ref, a quick answer: https://github.com/rsyslog/rsyslog-doc/issues/544#event-1459563817

**Pull request merged**